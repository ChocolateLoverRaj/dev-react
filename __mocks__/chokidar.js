// Mocker for chokidar
import EventEmitter from 'eventemitter3'
import sinon from 'sinon'

const chokidar = {}

export class FSWatcher extends EventEmitter { };

FSWatcher.prototype.close = sinon.fake(() => Promise.resolve())

chokidar.watch = sinon.fake(() => new FSWatcher())

export default chokidar
