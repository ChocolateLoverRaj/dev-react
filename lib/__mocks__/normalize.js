throw new Error('normalize is not setup yet')
export default 3