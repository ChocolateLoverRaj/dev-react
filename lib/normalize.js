// Like path.normalize, but removes trailing slashes and backslashes
import { normalize } from 'path'

const trailingSlashRegex = /[\\\/]$/

export default s => normalize(s).replace(trailingSlashRegex, '')
