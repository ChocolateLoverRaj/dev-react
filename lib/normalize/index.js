// Like path.normalize, but removes trailing (slashes and backslashes)
// Also converts all backslashes to slashes
const backslashRegex = /\\/g
const multipleSlashRegex = /\/+/g
const dotSlashRegex = /\/\.\//g
const dotDotSlashRegex = /[^/]*\/\.{2}\//g
const trailingSlashRegex = /\/$/

export default p => p
  .replace(backslashRegex, '/')
  .replace(multipleSlashRegex, '/')
  .replace(dotSlashRegex, '/')
  .replace(dotDotSlashRegex, '')
  .replace(trailingSlashRegex, '')
