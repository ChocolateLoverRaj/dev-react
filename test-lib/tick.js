// Returns a promise that gets resolved after process.nextTick()

const tick = () => new Promise(resolve => {
  process.nextTick(resolve)
})

module.exports = tick
