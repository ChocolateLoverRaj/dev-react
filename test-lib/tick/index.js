// Returns a promise that gets resolved after process.nextTick()

const tick = () => new Promise(resolve => {
  process.nextTick(resolve)
})

export default tick
