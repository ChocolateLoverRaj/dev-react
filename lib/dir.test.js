import Dir from './dir.js'
import normalize from './normalize.js'
import tick from '../test-lib/tick.js'
import chokidar, { _reset, _failClose } from '../__mocks__/chokidar.js'
import sinon from 'sinon'

const fakeOutputDir = (rejectCreateDir = false, rejectRemoveDir = false) => ({
  createDir: sinon.spy(() => rejectCreateDir ? Promise.reject() : Promise.resolve()),
  removeDir: sinon.spy(() => rejectRemoveDir ? Promise.reject() : Promise.resolve())
})

afterEach(_reset)

describe('constructor', () => {
  test('initial task', () => {
    const dir = new Dir(fakeOutputDir(), 'pages', 'page')
    expect(dir.task.getLines()).toMatchSnapshot()
  })

  describe('createDir', () => {
    test('success', async () => {
      const fake = fakeOutputDir()
      const dir = new Dir(fake, 'pages', 'page')
      let updateLines
      dir.task.emitter.on('update', () => {
        updateLines = dir.task.getLines()
      })
      expect(fake.createDir.calledOnceWith('page')).toBe(true)
      await fake.createDir.firstCall.returnValue
      await tick()
      const newLines = dir.task.getLines()
      expect(updateLines).toStrictEqual(newLines)
      expect(newLines).toMatchSnapshot()
    })

    test('fail', async () => {
      const fake = fakeOutputDir(true)
      const dir = new Dir(fake, 'pages', 'page')
      let updateLines
      dir.task.emitter.on('update', () => {
        updateLines = dir.task.getLines()
      })
      expect(fake.createDir.calledOnceWith('page')).toBe(true)
      await fake.createDir.firstCall.returnValue.catch(() => { })
      await tick()
      const newLines = dir.task.getLines()
      expect(updateLines).toStrictEqual(newLines)
      expect(newLines).toMatchSnapshot()
    })
  })

  test('watches chokidar', () => {
    new Dir(fakeOutputDir(), 'pages', 'page')
    expect(chokidar.watch.calledOnce).toBe(true)
    expect(normalize(chokidar.watch.firstCall.firstArg)).toBe('pages/page')
  })
})

describe('destroy', () => {
  test('initial task', async () => {
    const dir = new Dir(fakeOutputDir(), 'pages', 'page')
    const destroy = dir.destroy()
    expect(dir.task.getLines()).toMatchSnapshot()
    await destroy
  })

  test('all success', async () => {
    const fake = fakeOutputDir()
    const dir = new Dir(fake, 'pages', 'page')
    let updateLines
    dir.task.emitter.on('update', () => {
      updateLines = dir.task.getLines()
    })
    await dir.destroy()
    expect(fake.removeDir.calledOnceWith('page')).toBe(true)
    const newLines = dir.task.getLines()
    expect(updateLines).toStrictEqual(newLines)
    expect(newLines).toMatchSnapshot()
  })

  test('removeDir fail', async () => {
    const fake = fakeOutputDir(undefined, true)
    const dir = new Dir(fake, 'pages', 'page')
    let updateLines
    dir.task.emitter.on('update', () => {
      updateLines = dir.task.getLines()
    })
    await dir.destroy().catch(() => { })
    expect(fake.removeDir.calledOnceWith('page')).toBe(true)
    const newLines = dir.task.getLines()
    expect(updateLines).toStrictEqual(newLines)
    expect(newLines).toMatchSnapshot()
  })

  test('close chokidar fail', async () => {
    _failClose()
    const fake = fakeOutputDir()
    const dir = new Dir(fake, 'pages', 'page')
    let updateLines
    dir.task.emitter.on('update', () => {
      updateLines = dir.task.getLines()
    })
    await dir.destroy()
    expect(fake.removeDir.calledOnceWith('page')).toBe(true)
    const newLines = dir.task.getLines()
    expect(updateLines).toStrictEqual(newLines)
    expect(newLines).toMatchSnapshot()
  })
})
