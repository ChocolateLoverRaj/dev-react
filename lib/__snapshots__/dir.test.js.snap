// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constructor createDir fail 1`] = `
Array [
  "[94m📁 /page[39m",
  "[91m  Create dir[39m",
]
`;

exports[`constructor createDir success 1`] = `
Array [
  "[94m📁 /page[39m",
  "[32m  Create dir[39m",
]
`;

exports[`constructor initial task 1`] = `
Array [
  "[94m📁 /page[39m",
  "[93m  Create dir[39m",
]
`;

exports[`destroy all success 1`] = `
Array [
  "[94m📁 /page[39m",
  "[32m  Remove dir[39m",
  "[32m  Close chokidar[39m",
]
`;

exports[`destroy close chokidar fail 1`] = `
Array [
  "[94m📁 /page[39m",
  "[32m  Remove dir[39m",
  "[91m  Close chokidar[39m",
]
`;

exports[`destroy initial task 1`] = `
Array [
  "[94m📁 /page[39m",
  "[93m  Remove dir[39m",
  "[93m  Close chokidar[39m",
]
`;

exports[`destroy removeDir fail 1`] = `
Array [
  "[94m📁 /page[39m",
  "[91m  Remove dir[39m",
  "[32m  Close chokidar[39m",
]
`;
